File: tests/algorithms/x/PHP/searches/tabu_search.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _len($x) {
    if ($x === null) { return 0; }
    if (is_array($x)) { return count($x); }
    if (is_string($x)) { return strlen($x); }
    return strlen(strval($x));
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function path_cost($path, $graph) {
  global $best, $first;
  $total = 0;
  $i = 0;
  while ($i < count($path) - 1) {
  $u = $path[$i];
  $v = $path[$i + 1];
  $total = $total + $graph[$u][$v];
  $i = $i + 1;
};
  return $total;
};
  function generate_first_solution($graph, $start) {
  global $best, $first;
  $path = [];
  $visiting = $start;
  $total = 0;
  while (count($path) < count($graph)) {
  $path = _append($path, $visiting);
  $best_node = '';
  $best_cost = 1000000;
  foreach (array_keys($graph[$visiting]) as $n) {
  if (!(in_array($n, $path)) && $graph[$visiting][$n] < $best_cost) {
  $best_cost = $graph[$visiting][$n];
  $best_node = $n;
}
};
  if ($best_node == '') {
  break;
}
  $total = $total + $best_cost;
  $visiting = $best_node;
};
  $path = _append($path, $start);
  $total = $total + $graph[$visiting][$start];
  return ['cost' => $total, 'path' => $path];
};
  function copy_path($path) {
  global $best, $first, $graph;
  $res = [];
  $i = 0;
  while ($i < count($path)) {
  $res = _append($res, $path[$i]);
  $i = $i + 1;
};
  return $res;
};
  function find_neighborhood($sol, $graph) {
  global $best, $first;
  $neighbors = [];
  $i = 1;
  while ($i < _len($sol['path']) - 1) {
  $j = 1;
  while ($j < _len($sol['path']) - 1) {
  if ($i != $j) {
  $new_path = copy_path($sol['path']);
  $tmp = $new_path[$i];
  $new_path[$i] = $new_path[$j];
  $new_path[$j] = $tmp;
  $cost = path_cost($new_path, $graph);
  $neighbors = _append($neighbors, ['path' => $new_path, 'cost' => $cost]);
}
  $j = $j + 1;
};
  $i = $i + 1;
};
  return $neighbors;
};
  function find_swap($a, $b) {
  global $best, $first, $graph;
  $i = 0;
  while ($i < count($a)) {
  if ($a[$i] != $b[$i]) {
  return ['a' => $a[$i], 'b' => $b[$i]];
}
  $i = $i + 1;
};
  return ['a' => '', 'b' => ''];
};
  function tabu_search($first, $graph, $iters, $size) {
  $solution = $first;
  $best = $first;
  $tabu = [];
  $count = 0;
  while ($count < $iters) {
  $neighborhood = find_neighborhood($solution, $graph);
  if (count($neighborhood) == 0) {
  break;
}
  $best_neighbor = $neighborhood[0];
  $best_move = find_swap($solution['path'], $best_neighbor['path']);
  $i = 1;
  while ($i < count($neighborhood)) {
  $cand = $neighborhood[$i];
  $move = find_swap($solution['path'], $cand['path']);
  $forbidden = false;
  $t = 0;
  while ($t < count($tabu)) {
  if (($tabu[$t]['a'] == $move['a'] && $tabu[$t]['b'] == $move['b']) || ($tabu[$t]['a'] == $move['b'] && $tabu[$t]['b'] == $move['a'])) {
  $forbidden = true;
}
  $t = $t + 1;
};
  if ($forbidden == false && $cand['cost'] < $best_neighbor['cost']) {
  $best_neighbor = $cand;
  $best_move = $move;
}
  $i = $i + 1;
};
  $solution = $best_neighbor;
  $tabu = _append($tabu, $best_move);
  if (count($tabu) > $size) {
  $new_tab = [];
  $j = 1;
  while ($j < count($tabu)) {
  $new_tab = _append($new_tab, $tabu[$j]);
  $j = $j + 1;
};
  $tabu = $new_tab;
}
  if ($solution['cost'] < $best['cost']) {
  $best = $solution;
}
  $count = $count + 1;
};
  return $best;
};
  $graph = ['a' => ['b' => 20, 'c' => 18, 'd' => 22, 'e' => 26], 'b' => ['a' => 20, 'c' => 10, 'd' => 11, 'e' => 12], 'c' => ['a' => 18, 'b' => 10, 'd' => 23, 'e' => 24], 'd' => ['a' => 22, 'b' => 11, 'c' => 23, 'e' => 40], 'e' => ['a' => 26, 'b' => 12, 'c' => 24, 'd' => 40]];
  $first = generate_first_solution($graph, 'a');
  $best = tabu_search($first, $graph, 4, 3);
  echo rtrim(_str($best['path'])), PHP_EOL;
  echo rtrim(_str($best['cost'])), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/transpiler/FS/searches/tabu_search.fs
// Generated 2025-08-13 12:32 +0700

exception Break
exception Continue

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _dictCreate<'K,'V when 'K : equality> (pairs:('K * 'V) list) : System.Collections.Generic.IDictionary<'K,'V> =
    let d = System.Collections.Generic.Dictionary<'K, 'V>()
    for (k, v) in pairs do
        d.[k] <- v
    upcast d
let _dictGet<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) : 'V =
    match d.TryGetValue(k) with
    | true, v -> v
    | _ -> Unchecked.defaultof<'V>
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let _arrset (arr:'a array) (i:int) (v:'a) : 'a array =
    let mutable a = arr
    if i >= a.Length then
        let na = Array.zeroCreate<'a> (i + 1)
        Array.blit a 0 na 0 a.Length
        a <- na
    a.[i] <- v
    a
let rec _str v =
    let s = sprintf "%A" v
    let s = if s.EndsWith(".0") then s.Substring(0, s.Length - 2) else s
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
type Solution = {
    mutable _path: string array
    mutable _cost: int
}
type Swap = {
    mutable _a: string
    mutable _b: string
}
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
open System.Collections.Generic

let rec path_cost (_path: string array) (graph: System.Collections.Generic.IDictionary<string, System.Collections.Generic.IDictionary<string, int>>) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable _path = _path
    let mutable graph = graph
    try
        let mutable total: int = 0
        let mutable i: int = 0
        while i < ((Seq.length (_path)) - 1) do
            let u: string = _idx _path (int i)
            let v: string = _idx _path (int (i + 1))
            total <- total + (_dictGet (_dictGet graph ((string (u)))) ((string (v))))
            i <- i + 1
        __ret <- total
        raise Return
        __ret
    with
        | Return -> __ret
and generate_first_solution (graph: System.Collections.Generic.IDictionary<string, System.Collections.Generic.IDictionary<string, int>>) (start: string) =
    let mutable __ret : Solution = Unchecked.defaultof<Solution>
    let mutable graph = graph
    let mutable start = start
    try
        let mutable _path: string array = Array.empty<string>
        let mutable visiting: string = start
        let mutable total: int = 0
        try
            while (Seq.length (_path)) < (Seq.length (graph)) do
                try
                    _path <- Array.append _path [|visiting|]
                    let mutable best_node: string = ""
                    let mutable best_cost: int = 1000000
                    for n in (_dictGet graph ((string (visiting)))).Keys do
                        if (not (Seq.contains n _path)) && ((_dictGet (_dictGet graph ((string (visiting)))) ((string (n)))) < best_cost) then
                            best_cost <- _dictGet (_dictGet graph ((string (visiting)))) ((string (n)))
                            best_node <- n
                    if best_node = "" then
                        raise Break
                    total <- total + best_cost
                    visiting <- best_node
                with
                | Continue -> ()
                | Break -> raise Break
        with
        | Break -> ()
        | Continue -> ()
        _path <- Array.append _path [|start|]
        total <- total + (_dictGet (_dictGet graph ((string (visiting)))) ((string (start))))
        __ret <- { _path = _path; _cost = total }
        raise Return
        __ret
    with
        | Return -> __ret
and copy_path (_path: string array) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable _path = _path
    try
        let mutable res: string array = Array.empty<string>
        let mutable i: int = 0
        while i < (Seq.length (_path)) do
            res <- Array.append res [|(_idx _path (int i))|]
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and find_neighborhood (sol: Solution) (graph: System.Collections.Generic.IDictionary<string, System.Collections.Generic.IDictionary<string, int>>) =
    let mutable __ret : Solution array = Unchecked.defaultof<Solution array>
    let mutable sol = sol
    let mutable graph = graph
    try
        let mutable neighbors: Solution array = Array.empty<Solution>
        let mutable i: int = 1
        while i < ((Seq.length (sol._path)) - 1) do
            let mutable j: int = 1
            while j < ((Seq.length (sol._path)) - 1) do
                if i <> j then
                    let mutable new_path: string array = copy_path (sol._path)
                    let tmp: string = _idx new_path (int i)
                    new_path.[i] <- _idx new_path (int j)
                    new_path.[j] <- tmp
                    let _cost: int = path_cost (new_path) (graph)
                    neighbors <- Array.append neighbors [|{ _path = new_path; _cost = _cost }|]
                j <- j + 1
            i <- i + 1
        __ret <- neighbors
        raise Return
        __ret
    with
        | Return -> __ret
and find_swap (_a: string array) (_b: string array) =
    let mutable __ret : Swap = Unchecked.defaultof<Swap>
    let mutable _a = _a
    let mutable _b = _b
    try
        let mutable i: int = 0
        while i < (Seq.length (_a)) do
            if (_idx _a (int i)) <> (_idx _b (int i)) then
                __ret <- { _a = _idx _a (int i); _b = _idx _b (int i) }
                raise Return
            i <- i + 1
        __ret <- { _a = ""; _b = "" }
        raise Return
        __ret
    with
        | Return -> __ret
and tabu_search (first: Solution) (graph: System.Collections.Generic.IDictionary<string, System.Collections.Generic.IDictionary<string, int>>) (iters: int) (size: int) =
    let mutable __ret : Solution = Unchecked.defaultof<Solution>
    let mutable first = first
    let mutable graph = graph
    let mutable iters = iters
    let mutable size = size
    try
        let mutable solution: Solution = first
        let mutable best: Solution = first
        let mutable tabu: Swap array = Array.empty<Swap>
        let mutable count: int = 0
        try
            while count < iters do
                try
                    let neighborhood: Solution array = find_neighborhood (solution) (graph)
                    if (Seq.length (neighborhood)) = 0 then
                        raise Break
                    let mutable best_neighbor: Solution = _idx neighborhood (int 0)
                    let mutable best_move: Swap = find_swap (solution._path) (best_neighbor._path)
                    let mutable i: int = 1
                    while i < (Seq.length (neighborhood)) do
                        let cand: Solution = _idx neighborhood (int i)
                        let move: Swap = find_swap (solution._path) (cand._path)
                        let mutable forbidden: bool = false
                        let mutable t: int = 0
                        while t < (Seq.length (tabu)) do
                            if ((((_idx tabu (int t))._a) = (move._a)) && (((_idx tabu (int t))._b) = (move._b))) || ((((_idx tabu (int t))._a) = (move._b)) && (((_idx tabu (int t))._b) = (move._a))) then
                                forbidden <- true
                            t <- t + 1
                        if (forbidden = false) && ((cand._cost) < (best_neighbor._cost)) then
                            best_neighbor <- cand
                            best_move <- move
                        i <- i + 1
                    solution <- best_neighbor
                    tabu <- Array.append tabu [|best_move|]
                    if (Seq.length (tabu)) > size then
                        let mutable new_tab: Swap array = Array.empty<Swap>
                        let mutable j: int = 1
                        while j < (Seq.length (tabu)) do
                            new_tab <- Array.append new_tab [|(_idx tabu (int j))|]
                            j <- j + 1
                        tabu <- new_tab
                    if (solution._cost) < (best._cost) then
                        best <- solution
                    count <- count + 1
                with
                | Continue -> ()
                | Break -> raise Break
        with
        | Break -> ()
        | Continue -> ()
        __ret <- best
        raise Return
        __ret
    with
        | Return -> __ret
let graph: System.Collections.Generic.IDictionary<string, System.Collections.Generic.IDictionary<string, int>> = _dictCreate [("a", _dictCreate [("b", 20); ("c", 18); ("d", 22); ("e", 26)]); ("b", _dictCreate [("a", 20); ("c", 10); ("d", 11); ("e", 12)]); ("c", _dictCreate [("a", 18); ("b", 10); ("d", 23); ("e", 24)]); ("d", _dictCreate [("a", 22); ("b", 11); ("c", 23); ("e", 40)]); ("e", _dictCreate [("a", 26); ("b", 12); ("c", 24); ("d", 40)])]
let first: Solution = generate_first_solution (graph) ("a")
let mutable best: Solution = tabu_search (first) (graph) (4) (3)
ignore (printfn "%s" (_str (best._path)))
ignore (printfn "%s" (_str (best._cost)))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
